soci/soci_index.go

```diff
@@ -94,6 +112,60 @@ type IndexDescriptorInfo struct {
 	CreatedAt time.Time
 }
 
+// DecodeIndex deserializes a JSON blob in an io.Reader
+// into a SOCI index. The blob can be either an OCI 1.1 Artifact
+// or an OCI 1.0 Manifest
+func DecodeIndex(r io.Reader, index *Index) error {
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	return UnmarshalIndex(b, index)
+}
+
+// UnmarshalIndex deserializes a JSON blob in a byte array
+// into a SOCI index. The blob can be either an OCI 1.1 Artifact
+// or an OCI 1.0 Manifest
+func UnmarshalIndex(b []byte, index *Index) error {
+	err := json.Unmarshal(b, index)
+	if err == nil && index.MediaType == ocispec.MediaTypeArtifactManifest {
+		return nil
+	}
+	var manifest ocispec.Manifest
+	err = json.Unmarshal(b, &manifest)
+	if err == nil {
+		fromManifest(manifest, index)
+		return nil
+	}
+	return err
+}
+
+// fromManifest converts and OCI 1.0 Manifest to a SOCI Index
+func fromManifest(manifest ocispec.Manifest, index *Index) {
+	index.MediaType = manifest.MediaType
+	index.ArtifactType = SociIndexArtifactType
+	index.Blobs = manifest.Layers
+	index.Subject = manifest.Subject
+	index.Annotations = manifest.Annotations
+}
+
+// Marshal serializes a SOCI index into a JSON blob.
+// The JSON blob will be either an OCI 1.1 Artifact or
+// an OCI 1.0 Manifest, depending on how the index was created.
+func (i *Index) Marshal() ([]byte, error) {
+	if i.MediaType == OCIArtifactManifestMediaType {
```

Comment:
maybe be specific in log or comment which paths are 1.1 and which are 1.0?  Honestly my preference would break this out to make it easier to add future types and be crystal clear.  Into something like :

```
func (i *Index) Marshal() ([]byte, error) {
	if i.MediaType == OCIArtifactManifestMediaType {
              return marshalOCI11(i)
        } else  {
              return marshalOCI10(i)
        }
```

but honestly I'm flexible